wayfire/ipc.py

```diff
@@ -69,6 +64,49 @@ def read_message(self):
             raise Exception(response["error"])
         return response
 
+    def send_json(self, msg, timeout=1):
+        if 'method' not in msg:
+            raise Exception("Malformed JSON request: missing method!")
+
+        data = js.dumps(msg)
+        header = len(data).to_bytes(4, byteorder="little")
+        self.client.send(header)
+        self.client.send(data)
+
+        end_time = time.time() + timeout
+        while True:
+            remaining_time = end_time - time.time()
+            if remaining_time <= 0:
+                raise Exception("Response timeout")
+
+            readable, _, _ = select.select([self.client], [], [], remaining_time)
+            if readable:
+                try:
+                    response = self.read_message()
+                except Exception as e:
+                    raise Exception(f"Error reading message: {e}")
+
+                if 'event' in response:
+                    self.pending_events.append(response)
+                    continue
+
+                return response
+            else:
+                # The connection is closed due to a timeout
+                self.connect_client(self.socket_name)
```

Comment:
this reconnect here seems very weird to me, we're raising the exception, the user should decide what to do

Reply:
# We have two options:

## The worse option, which will lead to issues later:
`    Exception("Response timeout, client connection closed")`

In this case, we would remove the line for connecting the client, which is not ideal. Every time a timeout occurs, the client-side code would need to handle socket = WayfireSocket() again.
The issue here is the excessive checking to see if the client is still connected, all due to a send_json timeout.
It's not part of the user side to keep the client connected after `socket = WayfireSocket()`

## The better option:
   Set self.reconnect_on_client_timeout = True in __init__

```
Exception("Response timeout, reconnecting client")
if self.reconnect_on_client_timeout:
    self.connect_client(self.socket_name)
```

This way, the user can set socket.reconnect_client = False if they don't want the client to automatically reconnect.

it's not strange btw, the message 'Response timeout' indicates that the request didn't happen